File: q0394/q0394.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        cur_num = cur_str = ""
        for i in s:
            if i == '[':
                stack.append(cur_str)
                stack.append(int(cur_num))
                cur_num = cur_str = ""
            elif i == ']':
                num = stack.pop()
                prev_str = stack.pop()
                cur_str = prev_str + cur_str * num
            elif i.isdigit():
                cur_num += i
            else:
                cur_str += i
        return cur_str

File: q0394/q0394.java
package q0394;
import java.util.Stack;

class Solution {
    public String decodeString(String s) {
        Stack<String> strStack = new Stack<>();
        Stack<Integer> intStack = new Stack<>();
        String currNum = "";
        String currStr = "";
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '[') {
                int num = currNum.equals("") ? 0 : Integer.parseInt(currNum);
                intStack.push(num);
                currNum = "";

                strStack.push(currStr);
                currStr = "";
            } else if (c == ']') {
                Integer num = intStack.pop();
                String prevStr = strStack.pop();
                currStr = prevStr + currStr.repeat(num);
            } else if (Character.isDigit(c)) {
                currNum += c;
            } else {
                currStr += c;
            }
        }
        return currStr;
    }
}